libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/manager/MetadataSyncManager.kt

```diff
@@ -0,0 +1,259 @@
+     *
+     * @author bhariharan
+     */
+    interface MetadataSyncCallback {
+        /**
+         * Callback triggered when metadata sync completes.
+         *
+         * @param metadata Metadata.
+         */
+        fun onSyncComplete(metadata: Metadata?)
+    }
+
+    companion object {
+        const val SOUP_NAME = "sfdcMetadata"
+        const val QUERY = "SELECT {" + SOUP_NAME + ":_soup} FROM {" + SOUP_NAME +
+                "} WHERE {" + SOUP_NAME + ":" + Constants.ID + "} = '%s'"
+        private const val TAG = "MetadataSyncManager"
+        private val INDEX_SPECS = arrayOf(
+            IndexSpec(Constants.ID, SmartStore.Type.json1)
+        )
+        private val INSTANCES: MutableMap<String, MetadataSyncManager> = HashMap()
+
+        /**
+         * Returns the instance of this class associated with current user.
+         *
+         * @return Instance of this class.
+         */
+        @Synchronized @JvmStatic
+        fun getInstance(): MetadataSyncManager {
+            return getInstance(null, null)
+        }
+
+        /**
+         * Returns the instance of this class associated with this user account.
+         *
+         * @param account User account.
+         * @return Instance of this class.
+         */
+        @Synchronized
+        fun getInstance(account: UserAccount?): MetadataSyncManager {
+            return getInstance(account, null)
+        }
+
+        /**
+         * Returns the instance of this class associated with this user and community.
+         *
+         * @param account User account.
+         * @param communityId Community ID.
+         * @return Instance of this class.
+         */
+        @Synchronized
+        fun getInstance(account: UserAccount?, communityId: String?): MetadataSyncManager {
+            return getInstance(account, communityId, null)
+        }
+
+        /**
+         * Returns the instance of this class associated with this user, community and SmartStore.
+         *
+         * @param account User account. Pass null to use current user.
+         * @param communityId Community ID. Pass null if not applicable.
+         * @param smartStore SmartStore instance. Pass null to use current user default SmartStore.
+         * @return Instance of this class.
+         */
+        @Synchronized
+        fun getInstance(
+            account: UserAccount?, communityId: String?,
+            smartStore: SmartStore?
+        ): MetadataSyncManager {
+            val user = account ?: MobileSyncSDKManager.getInstance().userAccountManager.cachedCurrentUser
+            val store = smartStore ?: MobileSyncSDKManager.getInstance().getSmartStore(user, communityId)
+            val syncManager: SyncManager =
+                SyncManager.getInstance(user, communityId, store)
+            val uniqueId = ((if (user != null) user.userId else "") + ":"
+                    + store.database.path)
+            var instance = INSTANCES[uniqueId]
+            if (instance == null) {
```

Comment:
Optional: This could Elvis with the `instance` initialization above.  A quick regex search across the module reviewing if.*==null can find these.